framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,125 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseTypes.h"
+#include <vector>
+
+template <typename T, bool is_ad>
+class IndexableProperty
+{
+public:
+  static InputParameters validParams();
+
+  IndexableProperty(T * host,
+                    const std::string & mat_prop_param = "mat_prop",
+                    const std::string & component_param = "component");
+
+  GenericReal<is_ad> operator[](int i) const;
+  operator bool() const;
+
+protected:
+  template <typename P>
+  void getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                         unsigned int components);
+
+  const GenericMaterialProperty<Real, is_ad> * _mat_prop_real;
+  const GenericMaterialProperty<RealVectorValue, is_ad> * _mat_prop_real_vector_value;
+  const GenericMaterialProperty<RankTwoTensor, is_ad> * _mat_prop_rank_two_tensor;
+  const GenericMaterialProperty<RankThreeTensor, is_ad> * _mat_prop_rank_three_tensor;
+  const GenericMaterialProperty<RankFourTensor, is_ad> * _mat_prop_rank_four_tensor;
+
+  T * _host;
+
+  const std::string & _mat_prop_param;
+  const std::string & _mat_prop_name;
+  const std::string & _component_param;
+
+  const std::vector<unsigned int> _component;
+};
+
+template <typename T, bool is_ad>
+InputParameters
+IndexableProperty<T, is_ad>::validParams()
+{
+  auto params = T::validParams();
+  params.template addRequiredParam<MaterialPropertyName>("mat_prop",
```

Comment:
Do we have a convention for the parameter name for a material property? It seems like `property` or something would be more consistent with other classes, but I'd need to look around for examples.

Reply:
We use add param  `mat_prop` 12 times and `property` 4 times

Reply:
Bummer. I like `property` better. It's actually shorter, and I'm in favor of avoiding abbreviations when possible.

Reply:
`mat_prop`
`property`

same length. I agree regarding abbreviations. I don't think I came up with `mat_prop`. If you want me to change it, weighing consistency vs. nicer parameter name, I'd be happy to do it.